backend/cfg/cfg_polling.ml

```diff
@@ -0,0 +1,316 @@
+    | Poll -> (Poll, instr.dbg) :: points)
+  | Reloadretaddr | Pushtrap _ | Poptrap | Prologue -> points
+
+let add_poll_alloc_or_calls_terminator :
+    Cfg.terminator Cfg.instruction -> polling_points -> polling_points =
+ fun term points ->
+  match term.desc with
+  | Never -> assert false
+  | Always _ | Parity_test _ | Truth_test _ | Float_test _ | Int_test _
+  | Switch _ | Return | Raise _ | Specific_can_raise _ ->
+    (* CR xclerc for xclerc: double check the `Specific_can_raise` case. *)
+    points
+  | Tailcall_self _ | Tailcall_func _ -> (Function_call, term.dbg) :: points
+  | Call _ -> (Function_call, term.dbg) :: points
+  | Call_no_return { alloc = false; func_symbol = _; ty_res = _; ty_args = _ }
+  | Prim
+      { op =
+          External { alloc = false; func_symbol = _; ty_res = _; ty_args = _ };
+        label_after = _
+      } ->
+    points
+  | Call_no_return { alloc = true; func_symbol = _; ty_res = _; ty_args = _ }
+  | Prim
+      { op = External { alloc = true; func_symbol = _; ty_res = _; ty_args = _ };
+        label_after = _
+      } ->
+    (External_call, term.dbg) :: points
+  | Prim { op = Alloc _; label_after = _ } -> (Alloc, term.dbg) :: points
+  | Prim { op = Checkbound _; label_after = _ } -> points
+  | Prim { op = Probe _; label_after = _ } -> points
+
+let find_poll_alloc_or_calls : Cfg.t -> polling_points =
+ fun cfg ->
+  Cfg.fold_blocks cfg ~init:[] ~f:(fun _label block acc ->
+      let acc =
+        DLL.fold_right ~init:acc ~f:add_poll_alloc_or_calls_basic block.body
+      in
+      let acc = add_poll_alloc_or_calls_terminator block.terminator acc in
+      acc)
+
+let contains_polls : Cfg.t -> bool =
+ fun cfg ->
+  Cfg.fold_blocks cfg ~init:false ~f:(fun _label block acc ->
+      acc
+      || DLL.exists block.body ~f:(fun (instr : Cfg.basic Cfg.instruction) ->
+             match[@ocaml.warning "-4"] instr.Cfg.desc with
+             | Cfg.Op Poll -> true
+             | _ -> false))
+
+let is_disabled fun_name =
+  (not Config.poll_insertion)
+  || !Flambda_backend_flags.disable_poll_insertion
+  || function_is_assumed_to_never_poll fun_name
+
+let instrument_fundecl :
+    future_funcnames:Misc.Stdlib.String.Set.t ->
+    Cfg_with_layout.t ->
+    Cfg_with_layout.t =
+ fun ~future_funcnames:_ cfg_with_layout ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  if is_disabled cfg.fun_name
+  then cfg_with_layout
+  else
+    let block_needs_poll = polled_loops_analysis cfg_with_layout in
+    (* CR xclerc for xclerc: use `Cfg_with_infos`, to cache the computations? *)
+    let doms = Cfg_dominators.build cfg in
+    let back_edges = Cfg_loop_infos.compute_back_edges cfg doms in
+    let added_poll =
+      instr_cfg_with_layout cfg_with_layout ~block_needs_poll ~back_edges
+    in
+    (match cfg.fun_poll with
+    | Error_poll -> (
+      match find_poll_alloc_or_calls cfg with
+      | [] -> ()
+      | poll_error_instrs -> raise (Error (Poll_error poll_error_instrs)))
```

Comment:
Sort the list by source location to print nicer user errors.

Reply:
Upstream does not sort, but I can see
how it would be ordered or maybe
almost ordered.